pkg/kafka/partition/reader.go

```diff
@@ -114,7 +154,251 @@ func (p *Reader) run(ctx context.Context) error {
+	if !ok {
+		return 0, errors.New("unexpected response type")
+	}
+	if expected, actual := 1, len(listRes.Topics); actual != expected {
+		return 0, fmt.Errorf("unexpected number of topics in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if expected, actual := p.kafkaCfg.Topic, listRes.Topics[0].Topic; expected != actual {
+		return 0, fmt.Errorf("unexpected topic in the response (expected: %s, got: %s)", expected, actual)
+	}
+	if expected, actual := 1, len(listRes.Topics[0].Partitions); actual != expected {
+		return 0, fmt.Errorf("unexpected number of partitions in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if expected, actual := p.partitionID, listRes.Topics[0].Partitions[0].Partition; actual != expected {
+		return 0, fmt.Errorf("unexpected partition in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if err := kerr.ErrorForCode(listRes.Topics[0].Partitions[0].ErrorCode); err != nil {
+		return 0, err
+	}
+
+	return listRes.Topics[0].Partitions[0].Offset, nil
+}
+
+// processNextFetchesUntilTargetOrMaxLagHonored process records from Kafka until at least the maxLag is honored.
+// This function does a best-effort to get lag below targetLag, but it's not guaranteed that it will be
+// reached once this function successfully returns (only maxLag is guaranteed).
+func (p *Reader) processNextFetchesUntilTargetOrMaxLagHonored(ctx context.Context, targetLag, maxLag time.Duration, recordsChan chan<- []Record) error {
+	logger := log.With(p.logger, "target_lag", targetLag, "max_lag", maxLag)
+	level.Info(logger).Log("msg", "partition reader is starting to consume partition until target and max consumer lag is honored")
+
+	attempts := []func() (time.Duration, error){
+		// First process fetches until at least the max lag is honored.
+		func() (time.Duration, error) {
+			return p.processNextFetchesUntilLagHonored(ctx, maxLag, logger, recordsChan)
+		},
+
+		// If the target lag hasn't been reached with the first attempt (which stops once at least the max lag
+		// is honored) then we try to reach the (lower) target lag within a fixed time (best-effort).
+		// The timeout is equal to the max lag. This is done because we expect at least a 2x replay speed
+		// from Kafka (which means at most it takes 1s to ingest 2s of data): assuming new data is continuously
+		// written to the partition, we give the reader maxLag time to replay the backlog + ingest the new data
+		// written in the meanwhile.
+		func() (time.Duration, error) {
+			timedCtx, cancel := context.WithTimeoutCause(ctx, maxLag, errWaitTargetLagDeadlineExceeded)
+			defer cancel()
+
+			return p.processNextFetchesUntilLagHonored(timedCtx, targetLag, logger, recordsChan)
+		},
+
+		// If the target lag hasn't been reached with the previous attempt that we'll move on. However,
+		// we still need to guarantee that in the meanwhile the lag didn't increase and max lag is still honored.
+		func() (time.Duration, error) {
```

Comment:
Bizarre that after attempting to catch up to `targetLag`, we could find ourselves back at `maxLag` again. I'm keen to see `targetLag` removed in future if we can show it doesn't add much 😄

Reply:
Agreed. As discussed offline, let's run this and see how it performs in reality then loop back and simplify by removing the targetLag if its not a problem for Loki.